src/blob/index/bptree/core.rs

```diff
@@ -0,0 +1,283 @@
+/// structure of b+-tree index file from the beginning:
+/// 1. Header
+/// 2. Bloom filter
+/// 3. Sorted by keys array of record headers (serialized size of one record header is 61 bytes)
+/// 4. Metadata
+/// 5. Sorted by keys array of pairs `(key, offset of record header in file)` (these are leaf nodes
+///    of our B+ tree; as opposed to using record headers as leaves, this approach allows to include
+///    more elements in one block and reducing height of the tree (but increasing on one block
+///    that we need to read); this also will be useful for scanning queries: instead of scanning
+///    61 bytes headers (in case of 4-bytes key) we scan 12 bytes pairs (in case of 4-bytes key and
+///    8-bytes offset) which allows to read less bytes in 5 times)
+/// 6. Non-leaf nodes of btree in format
+///    `NodeMeta | keys_arr | pointers_arr`,
+///    where pointer is offset in file of underlying node with searched key
+use super::prelude::*;
+
+pub(super) const BLOCK_SIZE: usize = 4096;
+
+#[derive(Debug, Clone)]
+pub(crate) struct BPTreeFileIndex {
+    file: File,
+    header: IndexHeader,
+    metadata: TreeMeta,
+}
+
+#[async_trait::async_trait]
+impl FileIndexTrait for BPTreeFileIndex {
+    async fn from_file(name: FileName, ioring: Option<Rio>) -> Result<Self> {
+        trace!("open index file");
+        let file = File::open(name.to_path(), ioring)
+            .await
+            .context(format!("failed to open index file: {}", name))?;
+        let header = Self::read_index_header(&file).await?;
+        let metadata = Self::read_tree_meta(&file, &header).await?;
+
+        Ok(Self {
+            file,
+            header,
+            metadata,
+        })
+    }
+
+    async fn from_records(
+        path: &Path,
+        ioring: Option<Rio>,
+        headers: &InMemoryIndex,
+        filter: &Bloom,
+        recreate_index_file: bool,
+    ) -> Result<Self> {
+        let res = Self::serialize(headers, filter)?;
+        let (mut header, metadata, buf) = res;
+        clean_file(path, recreate_index_file)?;
+        let file = File::create(path, ioring)
```

Comment:
Actually, I don't create file 2 times (if file exists, then `clean_file` will truncate it, `File::create` will return descriptor for existing file). In `clean_file` I changed file deletion on truncation, because it's faster. Then I just recreate descriptor with our own file implemenation. If I'll just use `write_at(0, ...)`, seems like, I may get wrong bytes in tail.
But I missed case, when file doesn't exists in `clean_file`, changed that

Reply:
This is not critical, because dumping indexes in itself is a rare operation, and dumping indexes into an existing file is even rarer (occurs only in some exceptional cases)